graphjet-core/src/main/java/com/twitter/graphjet/algorithms/counting/user/TopSecondDegreeByCountUserRecsGenerator.java

```diff
@@ -0,0 +1,110 @@
+/**
+ * Copyright 2016 Twitter. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.twitter.graphjet.algorithms.counting.user;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.PriorityQueue;
+
+import com.google.common.collect.Lists;
+import com.twitter.graphjet.algorithms.*;
+import com.twitter.graphjet.algorithms.RecommendationInfo;
+import com.twitter.graphjet.algorithms.counting.GeneratorHelper;
+import com.twitter.graphjet.hashing.SmallArrayBasedLongToDoubleMap;
+import it.unimi.dsi.fastutil.longs.LongList;
+
+public class TopSecondDegreeByCountUserRecsGenerator {
+
+  /**
+   * Generate a list of recommendations based on given list of candidate nodes and the original request
+   * @param request         original request message, contains filtering criteria
+   * @param candidateNodes  list of candidate nodes
+   * @return                list of {@link UserRecommendationInfo}
+   */
+  public static List<RecommendationInfo> generateUserRecs(
+    TopSecondDegreeByCountRequestForUser request,
+    List<NodeInfo> candidateNodes) {
+
+    int maxNumResults = Math.min(request.getMaxNumResults(), RecommendationRequest.MAX_RECOMMENDATION_RESULTS);
+
+    PriorityQueue<NodeInfo> qualifiedNodes =
+      getQualifiedNodes(candidateNodes, request.getMinUserPerSocialProof(), maxNumResults);
+
+    return getRecommendationsFromNodes(request, qualifiedNodes);
+  }
+
+  private static PriorityQueue<NodeInfo> getQualifiedNodes(
+    List<NodeInfo> nodeInfoList,
+    Map<Byte, Integer> minSocialProofSizes,
+    int maxNumResults) {
+    PriorityQueue<NodeInfo> topResults = new PriorityQueue<>(maxNumResults);
+
+    for (NodeInfo nodeInfo : nodeInfoList) {
+      if (isQualifiedSocialProof(minSocialProofSizes, nodeInfo.getSocialProofs())) {
+        GeneratorHelper.addResultToPriorityQueue(topResults, nodeInfo, maxNumResults);
+      }
+    }
+    return topResults;
+  }
+
+  private static boolean isQualifiedSocialProof(
```

Comment:
in recommendation team's repositories, we prefer to user the term Valid instead of Qualified to refer to those candidates passing all filters.